File: frontend/src/api/match.ts
import api from './index'

// 匹配分析请求接口
export interface MatchAnalysisRequest {
  resume_id: number
  job_id: number
}

// 差距项接口
export interface GapItem {
  type: string
  skill_name?: string
  current_level?: number
  target_level?: number
  current_years?: number
  target_years?: number
  current_education?: string
  target_education?: string
  experience_type?: string
  severity: 'low' | 'medium' | 'high'
  description: string
}

// 建议接口
export interface Recommendation {
  type: string
  description: string
  priority: 'low' | 'medium' | 'high'
  estimated_time?: string
}

// 匹配分析结果接口
export interface MatchAnalysisResult {
  id: string
  user_id: number
  job_id: number
  position_title: string
  skill_match: number
  experience_match: number
  education_match: number
  overall_score: number
  summary?: string
  company?: string | null
  city?: string | null
  detail_json?: { analysis_source?: string; rule_summary?: string } | null
  gaps: GapItem[]
  recommendations: Recommendation[]
  created_at: string
  analyzed_at: string
}

// 匹配列表项接口
export interface MatchListItem {
  id: string
  job_id: number
  position_title: string
  overall_score: number
  skill_match: number
  experience_match: number
  education_match: number
  created_at: string
  analyzed_at: string
}

// 学习任务学习内容（知识点来自管理员知识库检索，练习题首次生成后缓存）
export interface StudyKnowledgeItem {
  content: string
  doc_title: string
  similarity: number
}

export interface StudyQuestionItem {
  question: string
  reference_answer: string
}

export interface StudyAnswerFeedback {
  answer: string
  feedback: string
  score?: number | null
}

export interface TaskStudy {
  task_id: number
  task_name: string
  knowledge: StudyKnowledgeItem[]
  questions: StudyQuestionItem[]
  answers: Record<string, StudyAnswerFeedback>
  batch: number
  total_generated: number
  source: 'cache' | 'llm' | 'rag_only'
}

// 仪表盘学习进度（用户最新一份学习计划的真实完成度）
export interface ProgressTaskItem {
  id: number
  task_name: string
  status: 'todo' | 'in_progress' | 'done'
  priority: 'low' | 'medium' | 'high'
}

export interface LearningProgress {
  has_plan: boolean
  position_title?: string | null
  total_tasks: number
  done_tasks: number
  in_progress_tasks: number
  progress: number
  tasks: ProgressTaskItem[]
}

// 学习任务接口
export interface LearningTask {
  id: number
  plan_id: number
  task_name: string
  description?: string
  resource_url?: string
  priority: 'low' | 'medium' | 'high'
  estimated_days?: number
  status: 'todo' | 'in_progress' | 'done'
  due_date?: string
}

// 学习计划接口
export interface LearningPlan {
  id: number
  user_id: number
  report_id: string
  content_json?: { tasks?: Array<Record<string, unknown>>; source?: string }
  status: string
  tasks: LearningTask[]
}

// 分析匹配度
export function analyzeMatch(data: MatchAnalysisRequest): Promise<MatchAnalysisResult> {
  // 完整匹配含 LLM 学习计划生成（实测 1-2 分钟），放宽超时到 5 分钟
  return api.post<MatchAnalysisResult>('/match/analyze', data, { timeout: 300000 }).then(r => r.data)
}

// 获取匹配列表
export function getMatches(): Promise<MatchListItem[]> {
  return api.get('/match/')
    .then(response => response.data.reports || [])
    .catch(error => {
      console.error('获取匹配结果失败:', error)
      return []
    })
}

// 获取匹配详情
export function getMatchDetail(matchId: string): Promise<MatchAnalysisResult> {
  return api.get<MatchAnalysisResult>(`/match/${matchId}`).then(r => r.data)
}

// 获取匹配报告对应的学习计划（含任务列表）
export function getMatchPlan(matchId: string): Promise<LearningPlan> {
  return api.get<LearningPlan>(`/match/${matchId}/plan`).then(r => r.data)
}

// 更新学习任务状态（todo: 待开始 / in_progress: 进行中 / done: 已完成）
export function updateTaskStatus(taskId: number, status: LearningTask['status']): Promise<LearningTask> {
  return api.patch<LearningTask>(`/match/plan/tasks/${taskId}`, { status }).then(r => r.data)
}

// 学习内容相关接口需现场调用 LLM（出题/点评耗时 10~70s），超时放宽到 3 分钟
const LLM_TIMEOUT = 180000

// 获取学习任务的学习内容（关联知识点 + 当前一批练习题 + 作答记录；无缓存题目时会现场生成）
export function getTaskStudy(taskId: number): Promise<TaskStudy> {
  return api.get<TaskStudy>(`/match/plan/tasks/${taskId}/study`, { timeout: LLM_TIMEOUT }).then(r => r.data)
}

// 换一批新题（旧题并入历史，新题不与历史重复，作答记录清空）
export function refreshTaskStudy(taskId: number): Promise<TaskStudy> {
  return api.post<TaskStudy>(`/match/plan/tasks/${taskId}/study/refresh`, null, { timeout: LLM_TIMEOUT }).then(r => r.data)
}

// 提交练习题作答，返回 AI 点评（得分 + 反馈）
export function submitTaskAnswer(
  taskId: number,
  question: string,
  answer: string
): Promise<{ question: string; answer: string; feedback: string; score?: number | null }> {
  return api
    .post(`/match/plan/tasks/${taskId}/study/answer`, { question, answer }, { timeout: LLM_TIMEOUT })
    .then(r => r.data)
}

// 获取仪表盘学习进度（最新学习计划的真实完成度）
export function getLearningProgress(): Promise<LearningProgress> {
  return api.get<LearningProgress>('/match/progress').then(r => r.data)
}

// 创建匹配（保留旧接口）
export function createMatch(data: { job_id: string; resume_data: any }) {
  return api.post('/match/create', data)
}

File: backend/app/schemas/match.py
from datetime import datetime
from enum import Enum
from typing import List, Optional
from pydantic import BaseModel, Field

from app.schemas.common import ORMBase


class GapSeverity(str, Enum):
    """差距严重程度"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class GapItem(BaseModel):
    """差距项"""
    type: str = Field(..., description="差距类型：skill, experience, education")
    skill_name: Optional[str] = Field(None, description="技能名称")
    current_level: Optional[int] = Field(None, description="当前水平")
    target_level: Optional[int] = Field(None, description="目标水平")
    current_years: Optional[float] = Field(None, description="当前年限")
    target_years: Optional[int] = Field(None, description="目标年限")
    current_education: Optional[str] = Field(None, description="当前学历")
    target_education: Optional[str] = Field(None, description="目标学历")
    experience_type: Optional[str] = Field(None, description="经验类型")
    severity: GapSeverity = Field(GapSeverity.MEDIUM, description="严重程度")
    description: str = Field(..., description="差距描述")


class Recommendation(BaseModel):
    """建议"""
    type: str = Field(..., description="建议类型")
    description: str = Field(..., description="建议描述")
    priority: GapSeverity = Field(GapSeverity.MEDIUM, description="优先级")
    estimated_time: Optional[str] = Field(None, description="预估时间")


class MatchCreate(BaseModel):
    """创建匹配请求"""
    resume_id: int = Field(..., description="简历ID")
    job_id: int = Field(..., description="岗位ID")


class MatchOut(ORMBase):
    """匹配报告输出"""
    id: str
    user_id: int
    job_id: int
    position_title: str
    skill_match: float = Field(..., ge=0, le=100, description="技能匹配度")
    experience_match: float = Field(..., ge=0, le=100, description="经验匹配度")
    education_match: float = Field(..., ge=0, le=100, description="教育匹配度")
    overall_score: float = Field(..., ge=0, le=100, description="综合评分")
    summary: Optional[str] = Field(None, description="LLM 综合分析文字（失败降级为规则文本）")
    detail_json: Optional[dict] = Field(None, description="分析来源与规则摘要等附加数据")
    company: Optional[str] = Field(None, description="目标岗位公司（来自岗位 parsed_json）")
    city: Optional[str] = Field(None, description="目标岗位城市（来自岗位 parsed_json）")
    analysis_source: Optional[str] = Field(None, description="综合分析来源：llm / rule")
    gaps: List[GapItem] = Field(default_factory=list, description="差距项")
    recommendations: List[Recommendation] = Field(default_factory=list, description="建议")
    learning_plan: List["LearningPlanOut"] = Field(
        default_factory=list, description="学习计划列表（含任务）"
    )
    created_at: datetime
    analyzed_at: datetime


class MatchListItem(BaseModel):
    """匹配报告列表项"""
    id: str
    job_id: int
    position_title: str
    overall_score: float
    skill_match: float
    experience_match: float
    education_match: float
    created_at: datetime
    analyzed_at: datetime


class MatchAnalysisResult(BaseModel):
    """匹配分析结果"""
    match_report: MatchOut
    feedback: str
    learning_plan: List[dict]


class LearningTaskOut(ORMBase):
    """学习任务输出"""
    id: int
    plan_id: int
    task_name: str
    description: Optional[str] = None
    resource_url: Optional[str] = None
    priority: str = "medium"
    estimated_days: Optional[int] = None
    status: str = "todo"
    due_date: Optional[str] = None


class LearningPlanOut(ORMBase):
    """学习计划输出"""
    id: int
    user_id: int
    report_id: str
    content_json: Optional[dict] = None
    status: str = "pending"
    tasks: List[LearningTaskOut] = Field(default_factory=list, description="学习任务列表")


class TaskStatus(str, Enum):
    """学习任务状态（需求 UC-012：待开始/进行中/已完成）"""
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    DONE = "done"


class TaskStatusUpdate(BaseModel):
    """学习任务状态更新请求"""
    status: TaskStatus = Field(..., description="任务状态：todo / in_progress / done")


class TaskAnswerIn(BaseModel):
    """练习题作答请求"""
    question: str = Field(..., min_length=1, description="题目文本（需为当前批练习题之一）")
    answer: str = Field(..., min_length=1, description="用户作答内容")


class ProgressTaskItem(BaseModel):
    """仪表盘学习进度任务项"""
    id: int
    task_name: str
    status: str = "todo"
    priority: str = "medium"


class LearningProgressOut(BaseModel):
    """仪表盘学习进度（取用户最新一份学习计划）"""
    has_plan: bool = Field(False, description="是否存在学习计划")
    position_title: Optional[str] = Field(None, description="对应匹配报告的岗位名称")
    total_tasks: int = Field(0, description="任务总数")
    done_tasks: int = Field(0, description="已完成任务数")
    in_progress_tasks: int = Field(0, description="进行中任务数")
    progress: int = Field(0, ge=0, le=100, description="完成进度百分比")
    tasks: List[ProgressTaskItem] = Field(default_factory=list, description="任务列表（按优先级）")


class StudyKnowledgeItem(BaseModel):
    """学习内容关联的知识点（来自平台管理员知识库 RAG 检索）"""
    content: str = Field(..., description="知识片段内容")
    doc_title: str = Field(..., description="来源文档标题")
    similarity: float = Field(..., description="与任务的相关度")


class StudyAnswerFeedback(BaseModel):
    """练习题作答记录与 AI 点评"""
    answer: str = Field("", description="用户作答内容")
    feedback: str = Field("", description="AI 点评")
    score: Optional[int] = Field(None, description="得分（0-100）")


class StudyQuestionItem(BaseModel):
    """学习任务练习题（LLM 生成，含参考答案）"""
    question: str = Field(..., description="练习题目")
    reference_answer: str = Field("", description="参考答案")


class TaskStudyOut(BaseModel):
    """学习任务的学习内容"""
    task_id: int
    task_name: str
    knowledge: List[StudyKnowledgeItem] = Field(
        default_factory=list, description="关联知识点（管理员知识库检索）"
    )
    questions: List[StudyQuestionItem] = Field(
        default_factory=list, description="当前一批练习题（首次生成后缓存）"
    )
    answers: dict[str, StudyAnswerFeedback] = Field(
        default_factory=dict, description="作答记录（题目文本 → 作答与点评）"
    )
    batch: int = Field(1, description="当前第几批题目（换一批后递增）")
    total_generated: int = Field(0, description="累计生成过的题目总数")
    source: str = Field("rag_only", description="题目来源：cache（缓存）/ llm（新生成）/ rag_only（仅知识点）")
